### ISSUE TITLE
AUTO_TITLE functionality breaks in tmux

### ISSUE THREAD

**Issue by gotyaoi** [robbyrussell/oh-my-zsh#3579]
In my tmux config, I have it set up to handle the window title like so:

```
set-option -g set-titles on
set-option -g set-titles-string '#{pane_title}'
set-option -g default-terminal 'screen-256color'
```

That is, get the title from the active pane's title. The function in oh my zsh for setting the window title looks like this:

```
function title {
   if [[ "$DISABLE_AUTO_TITLE" == "true" ]] || [[ "$EMACS" == *term* ]]; then
     return
   fi
   if [[ "$TERM" == screen* ]]; then
     print -Pn "\ek$1:q\e\\" #set screen hardstatus, usually truncated at 20 chars
   elif [[ "$TERM" == xterm* ]] || [[ $TERM == rxvt* ]] || [[ $TERM == ansi ]] || [[ "$TERM_PROGRAM" == "iTerm.app" ]]; then
     print -Pn "\e]2;$2:q\a" #set window name
     print -Pn "\e]1;$1:q\a" #set icon (=tab) name
   fi
 }
```

since tmux advises using `screen` or `screen-256color` for your TERM variable, this causes the first if condition to match, and the screen hardstatus doesn't seem to work for tmux, leading to a garbled or unchanging title. For tmux, I think  the statements in the elif branch makes sense, since then things just act like any other shell, which tmux can easily to hook into. Since tmux sets the $TMUX variable, that would probably be an easy way to detect this case, perhaps something like:

```
function title {
   if [[ "$DISABLE_AUTO_TITLE" == "true" ]] || [[ "$EMACS" == *term* ]]; then
     return
   fi
   if [[ "$TERM" == screen* ]] && [[ -z "$TMUX" ]]; then
     print -Pn "\ek$1:q\e\\" #set screen hardstatus, usually truncated at 20 chars
   elif [[ "$TERM" == xterm* ]] || [[ $TERM == rxvt* ]] || [[ $TERM == ansi ]] || [[ "$TERM_PROGRAM" == "iTerm.app" ]] || [[ -n "$TMUX" ]]; then
     print -Pn "\e]2;$2:q\a" #set window name
     print -Pn "\e]1;$1:q\a" #set icon (=tab) name
   fi
 }
```

**Comment by mcornella** [comment]
Hi @gotyaoi, first of all the title functionality has been updated and it's no longer the one you show. That said, I believe the same problem exists. 
For the time being you can set `DISABLE_AUTO_TITLE=true` in your zshrc file and, from what I gather of your tmux configuration, it will be set automatically by tmux (correct me if I'm wrong). But I don't see anything wrong with the extra check you posted, provided all tmux versions and configurations need to use the second branch of the if statement. The `[[ -n "$TMUX" ]]` is redundant though. 

If you feel comfortable doing so, please provide a pull request for quick review and let's see if we can get some testers on this. Otherwise I'll get on it in the next week.

Thanks!

**Comment by gotyaoi** [comment]
Ah, my apologies, was going with what was on disk.

If I set DISABLE_AUTO_TITLE, the title is set to the hostname initially, then not messed with further by oh my zsh, looks like.

I think the elif branch's behavior makes more sense than doing nothing by default, which is also why the -n test is necessary, I think. Without the -n, since $TERM is usually screen/screen-256color, it wouldn't match the second branch, right? Unless I'm misunderstanding the conditions...

Pending those questions, I can definitely submit a pull request.

**Comment by mcornella** [comment]
Yeah, you're definitely right about the `-n`, sorry :sweat_smile:
You can go ahead and submit a PR

**Comment by gotyaoi** [comment]
Ah, it just occurred to me that it may not have notified simply by referencing this issue in the pull request, so replying now if that's the case. The pull request is at https://github.com/robbyrussell/oh-my-zsh/issues/3599.

**Comment by gotyaoi** [comment]
So, I've closed the pull request for now as it seems that this is slightly more complicated than I thought. It turns out tmux can use the \ek...\e\ sequence fairly easily, it's just dependent on the `allow-rename` setting being on. In addition, tmux has `automatic-rename` and `automatic-rename-format`, which give a fairly fine grained control over the string that would be set by the escape sequence. The `automatic-rename` functionality appears to work regardless of the setting of `allow-rename`, but if the escape sequence is used automatic renaming stops.

Separate from this is the `set_titles*` settings I was using to control the title of my terminal.

Basically, the problem is, there's no objectively better way to handle things. Some users might want the escape sequence to be used, and completely ignore the window title. Some might only care about the window title and not want the escape sequence at all. Some might want both or neither, for their particular setup. This suggests config settings in .zshrc or somewhere, but I'm hesitant to say "shove a bunch of tmux specific logic into termsupport.zsh" because then there's some duplication between certain blocks and....

So, I guess I'm asking what do other people think about this situation/how it can be addressed.

**Comment by gotyaoi** [comment]
Ok, how does this look? It preserves the current behavior, but allows the user to set ZSH_THEME_TERM_TITLE_TYPE={screen|screenlong|xterm|both|bothlong} to get different behavior. The screenlong and bothlong ones are not entirely necessary, I just include them here for completeness.
```
# See: http://www.faqs.org/docs/Linux-mini/Xterm-Title.html#ss3.1
# Fully supports screen, iterm, and probably most modern xterm and rxvt
# (In screen, only short_tab_title is used)
# Limited support for Apple Terminal (Terminal can't set window and tab separately)
function title {
  [[ "$EMACS" == *term* ]] && return

  # if $2 is unset use $1 as default
  # if it is set and empty, leave it as is
  : ${2=$1}

  case $ZSH_THEME_TERM_TITLE_TYPE in
    screen)
      print -Pn "\ek$1:q\e\\" #set screen title string, usually truncated at 20 chars
      ;;
    screenlong)
      print -Pn "\ek$2:q\e\\"
      ;;
    xterm)
      print -Pn "\e]2;$2:q\a" #set window name
      print -Pn "\e]1;$1:q\a" #set icon (=tab) name
      ;;
    both)
      print -Pn "\ek$1:q\e\\"
      print -Pn "\e]2;$2:q\a"
      print -Pn "\e]1;$1:q\a"
      ;;
    bothlong)
      print -Pn "\ek$2:q\e\\"
      print -Pn "\e]2;$2:q\a"
      print -Pn "\e]1;$1:q\a"
      ;;
  esac
}

if [[ -z $ZSH_THEME_TERM_TITLE_TYPE ]] then;
  if [[ "$TERM" == screen* ]]; then
    ZSH_THEME_TERM_TITLE_TYPE=screen
  elif [[ "$TERM" == xterm* ]] || [[ "$TERM" == rxvt* ]] || [[ "$TERM" == ansi ]] || [[ "$TERM_PROGRAM" == "iTerm.app" ]]; then
    ZSH_THEME_TERM_TITLE_TYPE=xterm
  fi
fi

ZSH_THEME_TERM_TAB_TITLE_IDLE="%15<..<%~%<<" #15 char left truncated PWD
ZSH_THEME_TERM_TITLE_IDLE="%n@%m: %~"```

**Comment by fcrozat** [comment]
Switching tmux to use `screen-256color-s` and the referenced pull request allows applications setting terminal title to properly work in tmux (such as vim). And it ensures AUTO_TITLE still works fine (once you set `set -g set-titles on` and `set -g set-titles-string "#{pane_title}"` in tmux.conf)

**Comment by mcornella** [comment]
In my system, with the settings that @fcrozat suggested (except using `tmux-256color`), and modifying https://github.com/robbyrussell/oh-my-zsh/blob/4ddb2e384ab0840b1d8a6d4c95770ef8a6c25fcc/lib/termsupport.zsh#L20-L23

to include `tmux*` I can get a proper window title.

We change the case statement to include it in the `screen*` section (#7622). For that to work, I have to change use `set -g set-titles-string "#W"`, but since the `title` function uses the argument `$1` I don't get the full command being run.

In any case, use:
```
set -g set-titles on
set -g set-titles-string "#{pane_title}"
set -g default-terminal "xterm-256color"
```
to get the same behavior as a non-tmux terminal, or use
```
set -g set-titles on
set -g set-titles-string "#W"
set -g default-terminal "tmux-256color"
```
to get a similar behavior, but you won't get the full command in the title when running one. Perhaps we could change `$1` to `$2` in the title function above.
